kotlinx-coroutines-core/common/src/selects/Select.kt

```diff
@@ -52,606 +97,696 @@ public interface SelectBuilder<in R> {
-            // here we are definitely not going to RETRY_ATOMIC, so
-            // we must finish preparation of another operation before attempting to reach decision to select
-            otherOp.finishPrepare()
-            val decision = otherOp.atomicOp.decide(null) // try decide for success of operation
-            val update: Any = if (decision == null) otherOp.desc else NOT_SELECTED
-            impl._state.compareAndSet(this, update)
-            return decision
-        }
-
-        override val atomicOp: AtomicOp<*>
-            get() = otherOp.atomicOp
+    /**
+     * Finds the clause with the corresponding [clause object][SelectClause.clauseObject].
+     * If the reference to the list of clauses is already cleared due to completion/cancellation,
+     * this function returns `null`
+     */
+    private fun findClause(clauseObject: Any) = clauses?.run {
+        find { it.clauseObject === clauseObject } ?: error("Clause with object $clauseObject is not found")
     }
 
-    override fun performAtomicTrySelect(desc: AtomicDesc): Any? =
-        AtomicSelectOp(this, desc).perform(null)
-
-    override fun toString(): String = "SelectInstance(state=${_state.value}, result=${_result.value})"
+    // ==============
+    // = COMPLETION =
+    // ==============
 
-    private class AtomicSelectOp(
-        @JvmField val impl: SelectBuilderImpl<*>,
-        @JvmField val desc: AtomicDesc
-    ) : AtomicOp<Any?>() {
-        // all select operations are totally ordered by their creating time using selectOpSequenceNumber
-        override val opSequence = selectOpSequenceNumber.next()
-
-        init {
-            desc.atomicOp = this
-        }
+    /**
+     * Completes this `select` operation after the internal result is provided
+     * via [SelectInstance.trySelect] or [SelectInstance.selectInRegistrationPhase].
+     * (1) First, this function applies the [ProcessResultFunction] of the selected clause
+     * to the internal result.
+     * (2) After that, the [clean-up procedure][cleanup]
+     * is called to remove this `select` instance from other clause objects, and
+     * make it possible to collect it by GC after this `select` finishes.
+     * (3) Finally, the user-specified block is invoked
+     * with the processed result as an argument.
+     */
+    private suspend fun complete(): R {
+        assert { isSelected }
+        // Get the selected clause.
+        @Suppress("UNCHECKED_CAST")
+        val selectedClause = state.value as ClauseData<R>
+        // Process the internal result.
+        val blockArgument = selectedClause.processResult(internalResult)
+        // Perform the clean-up before the user-specified block
+        // invocation to guarantee the absence of memory leaks.
+        cleanup(selectedClause)
+        // TAIL-CALL OPTIMIZATION: the `suspend` block
+        // is invoked at the very end.
+        return selectedClause.invokeBlock(blockArgument)
```

Comment:
There are two places where an exception can be thrown and thus stacktrace recovery _with outer continuation_ should be used: `processResult` and `invokeBlock`.

I remember we decided that the performance of the select _at this level_ does not really matter, so instead of trying to find a hacky workaround I suggest simply giving up TCO here (we're allocating a lot anyways) and use proper stacktrace recovery

Reply:
We use this code on the fast path, so am not sure that this is a good idea

Reply:
```
internal open suspend fun doSelect(): R =
        if (isSelected) complete()  // Fast path
        else doSelectSuspend()      // Slow path
```
Here, we invoke `complete()` on the fast path

Reply:
When we first discussed the review and you were asked about benchmarks, the answer was "performance of the select is not important", which is fine.
Now, when we are talking about debuggability and UX, the performance is important for some reason, which may be  also fine in isolation.

But these two reasonings together just don't add up, please stick with a consistent approach, and if you are choosing performance over user experience -- it's better have some numbers to support the decision about giving up user's convenience.

I've pushed `SelectBenchmark` that just measures the overhead of `select` and the experiment shows that losing tail-call in `doSelect` costs 5% of throughput _at best_, which is completely reasonable for stacktrace recovery, so please implement it

Reply:
Also, I've measured the old select. New one is 50% faster, which is really good for the implementation that do not care about performance and indicates that we are on the right track

Reply:
I found even better solution:
```
return if (!RECOVER_STACK_TRACES) {
            // TAIL-CALL OPTIMIZATION: the `suspend` block
            // is invoked at the very end.
            selectedClause.invokeBlock(blockArgument)
        } else {
            // In the debug mode, we need to recover
            // the stack-trace of possible exception.
            // Thus, the tail-call optimization cannot be applied.
            try {
                selectedClause.invokeBlock(blockArgument)
            } catch (e: Throwable) {
                throw recoverStackTrace(e)
            }
        }
```

Updated the code correspondingly.